### PR TITLE
Add CircleCI step to publish horizon docker image for every release

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### Summary

Add CircleCI step to publish horizon docker image for every release

### Goal and scope

The goal of this PR is to publish a horizon docker image for every release of horizon.

### Summary of changes

* Add `publish_horizon_docker_image` , a circle CI job which builds and pushes a horizon docker image
* invoke  `publish_horizon_docker_image` on all tags which match the following regular expression `^horizon-v.*`

Note we will need to add a docker hub login and password in the CircleCI build environment.

The Dockerfile used to build the horizon image is:
https://github.com/stellar/go/blob/master/services/horizon/docker/Dockerfile

### Known limitations & issues

N / A

### What shouldn't be reviewed

N / A

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -206,6 +206,23 @@ jobs:
...
+      - checkout
+      - setup_remote_docker
+      - run:
+          name: Build and Push Docker image
+          command: |
+            # CIRCLE_TAG will be prefixed by "horizon-v", here we build the horizon docker image and tag it with stellar/horizon:$VERSION
+            # where version is CIRCLE_TAG without the "horizon-v" prefiix
+            VERSION=$(echo $CIRCLE_TAG | sed -e "s/^horizon-v//")
+            docker build -f services/horizon/docker/Dockerfile -t stellar/horizon:$VERSION .
+            docker login -u "$DOCKER_USER" -u "$DOCKER_PASS"
```
Comment (bartekn) [inline .circleci/config.yml]:
```suggestion
            docker login -u "$DOCKER_USER" -p "$DOCKER_PASS"
```

**Comment by ire-and-curses** [comment]
I believe this should wait until we are in agreement about how we'd like to distribute Docker images. We're having an internal discussion (currently blocked on me) about single source/multi-channel distribution. We can use the new docker-compose stuff for dev so I think merging this can be delayed.

Does that sound right or am I misunderstanding something?

**Comment by tamirms** [comment]
> I believe this should wait until we are in agreement about how we'd like to distribute Docker images. We're having an internal discussion (currently blocked on me) about single source/multi-channel distribution. We can use the new docker-compose stuff for dev so I think merging this can be delayed.
> 
> Does that sound right or am I misunderstanding something?

yeah, that's right. I think we're going to set up an internal docker registry. Once the internal docker registry is up and running we can publish docker images there. And after we accumulate experience running horizon in kubernetes we can think about publishing a public docker image

**Comment by brahman81** [comment]
@tamirms We have an internal Docker registry available, feel free to sync with @jacekn to get details

File: .circleci/config.yml [hunk 2]
```diff
@@ -298,6 +298,23 @@ jobs:
...
+      - checkout
+      - setup_remote_docker
+      - run:
+          name: Build and Push Docker image
+          command: |
+            echo "$DOCKER_PASS" | docker login -u "$DOCKER_USER" --password-stdin
+            # CIRCLE_TAG will be prefixed by "horizon-v", here we build the horizon docker image and tag it with stellar/horizon:$VERSION
+            # where version is CIRCLE_TAG without the "horizon-v" prefix
+            VERSION=${CIRCLE_TAG#horizon-v}
+            docker build -f services/horizon/docker/Dockerfile -t stellar/horizon:$VERSION -t stellar/horizon:latest .
```
Comment (leighmcculloch) [inline .circleci/config.yml]: We might not want the `latest` tag here if we are releasing patch releases for older versions of horizon. If we are currently on `v1.6.1` and need to release `v1.5.3`, `v1.5.3` will become `latest`.

Comment (leighmcculloch) [reply .circleci/config.yml]: I think it would be safer and sufficient to make pushing latest a manual op.

**Pull request merged**